### DISCUSSION TITLE
Unable to export custom report into Excel format(.csv/.xlsx) from Mordern pos

### DISCUSSION THREAD

**Discussion by excelexport** [microsoft/Dynamics365Commerce.Solutions#186] [Q&A] [answered]
What should be data structure to pass for this **data.reportDataSet** ??

************ Here is the code *************
import { IExtensionCommandContext } from "PosApi/Extend/Views/AppBarCommands";
import * as ReportDetailsView from "PosApi/Extend/Views/ReportDetailsView";
import { ProxyEntities } from "PosApi/Entities";
import { ObjectExtensions, StringExtensions } from "PosApi/TypeExtensions";

export default class ExportToCSVCommand extends ReportDetailsView.ReportDetailsExtensionCommandBase {

    private _reportData: ProxyEntities.ReportDataSet = null;
    private _reportTitle: string = StringExtensions.EMPTY;
    private _reportId: string = StringExtensions.EMPTY;

    constructor(context: IExtensionCommandContext<ReportDetailsView.IReportDetailsToExtensionCommandMessageTypeMap>) {
        super(context);
        this.id = "sampleExportToCSVCommand";
        this.label = "ExportToCSV";
        this.extraClass = "iconPickup";

        this.reportDataLoadedDataHandler = (data: ReportDetailsView.ReportDataLoadedData): void => {
            this.canExecute = true;
            this._reportData = data.reportDataSet;
        };
    }

    protected init(state: ReportDetailsView.IReportDetailsExtensionCommandState): void {
        this.isVisible = true;
        this._reportId = state.reportId;
        this._reportTitle = state.reportTitle;
    }

    protected execute(): void {
        this.context.logger.logInformational("Report title: " + JSON.stringify(this._reportTitle));
        this.context.logger.logInformational("Report Id: " + JSON.stringify(this._reportId));
        this.context.logger.logInformational("Print report data: " + JSON.stringify(this._reportData));

        let headerInited: boolean = false;
        let csvRPTHeader: string = "";
        let csvRPTContent: string = "";

        this._reportData.Output.forEach((row: ProxyEntities.ReportRow) => {
            row.RowData.forEach((property: ProxyEntities.CommerceProperty) => {
                // Add header title
                if (!headerInited) {
                    csvRPTHeader += property.Key;
                }
                if (!ObjectExtensions.isNullOrUndefined(property.Value.BooleanValue)) {
                    csvRPTContent += property.Value.BooleanValue.toString();
                } else if (!ObjectExtensions.isNullOrUndefined(property.Value.StringValue)) {
                    csvRPTContent += property.Value.StringValue;
                } else if (!ObjectExtensions.isNullOrUndefined(property.Value.DateTimeOffsetValue)) {
                    csvRPTContent += property.Value.DateTimeOffsetValue.toString();
                } else if (!ObjectExtensions.isNullOrUndefined(property.Value.DecimalValue)) {
                    csvRPTContent += property.Value.DecimalValue.toString();
                } else if (!ObjectExtensions.isNullOrUndefined(property.Value.IntegerValue)) {
                    csvRPTContent += property.Value.IntegerValue.toString();
                } else if (!ObjectExtensions.isNullOrUndefined(property.Value.LongValue)) {
                    csvRPTContent += property.Value.LongValue.toString();
                } else if (!ObjectExtensions.isNullOrUndefined(property.Value.ByteValue)) {
                    csvRPTContent += property.Value.ByteValue.toString();
                }
                if (!headerInited) {
                    csvRPTHeader += ",";
                }
                csvRPTContent += ",";
            });

            // Header title will be only initialized once
            if (!headerInited) {
                csvRPTHeader = csvRPTHeader.substr(0, csvRPTHeader.length - 1);
                csvRPTHeader += "\n";
            }
            csvRPTContent = csvRPTContent.substr(0, csvRPTContent.length - 1);
            csvRPTContent += "\n";
            headerInited = true;
        });

        // Create anchor element and set download attribute
        const link = document.createElement("a");
        link.setAttribute("href",
            `data:text/csv;charset=utf-8,${encodeURIComponent(csvRPTHeader + csvRPTContent)}`);
        link.setAttribute("download", "data.csv");

        // Append link to body and trigger click event
        document.body.appendChild(link);
        link.click();

        // Remove link from body
        document.body.removeChild(link);
    }
}

**Comment by ViktorVu602** [comment]
Hi @madyke, may I ask if you can help to look into this issue? This is an extension for Retail SDK.
Thank you!

**Reply by madyke** [reply, marked as answer]
Hello, our team has created a work item to review this request, however since it is a Retail SDK specific question, we are unlikely to have time to assist. From my understanding, support has already provided a relevant sample for the Commerce SDK implementation, and as Retail SDK is being deprecated in October our blanket recommendation for all Retail SDK questions is to upgrade to Commerce SDK.
